Readme.md
# idb-batch

> Perform batch operation on IndexedDB

[![](https://saucelabs.com/browser-matrix/idb-batch.svg)](https://saucelabs.com/u/idb-batch)

[![](https://img.shields.io/npm/v/idb-batch.svg)](https://npmjs.org/package/idb-batch)
[![](https://img.shields.io/travis/treojs/idb-batch.svg)](https://travis-ci.org/treojs/idb-batch)
[![](http://img.shields.io/npm/dm/idb-batch.svg)](https://npmjs.org/package/idb-batch)

Create/update/remove objects from IndexedDB store in one transaction [without blocking the main thread](http://stackoverflow.com/questions/10471759/inserting-large-quantities-in-indexeddbs-objectstore-blocks-ui).
This module also manually validates unique indexes, fixing bugs in [WebKit](https://bugs.webkit.org/show_bug.cgi?id=149107)
and [IndexedDBShim](https://github.com/axemclion/IndexedDBShim/issues/56).

## Example

Using [idb-factory](https://github.com/treojs/idb-factory) and [ES2016 async/await syntax](https://jakearchibald.com/2014/es7-async-functions/).
Check [test.js](./test/index.js) for more examples.

```js
import batch from 'idb-batch'
import { open } from 'idb-factory'

// open IndexedDB database with 2 stores
const db = await open('mydb', 1, upgradeCallback)

// modify object store
await batch(db, 'magazines', [
  { type: 'add', key: 1, value: { name: 'M1', frequency: 12 } },
  { type: 'add', key: 2, value: { name: 'M2', frequency: 24 } },
  { type: 'add', key: 3, value: { name: 'M3', frequency: 6 } },
  { type: 'del', key: 4 },
]).then((result) => {
  console.log(result) // [1, 2, 3, undefined]
}).catch((err) => {
  console.error(err)
})

function upgradeCallback(e) {
  e.target.result.createObjectStore('books', { keyPath: 'id' })  
  e.target.result.createObjectStore('magazines')  
}
```

### transactionalBatch(tr: IDBDatabase|IDBTransaction, storeOpsArr, opts = { parallel: false, extraStores: [] })

To-do

### batch(db: IDBDatabase|IDBTransaction, storeName: String, ops: Array|Object, opts: { parallel: false, extraStores: [] })

This creates a `readwrite` transaction to `storeName`, and performs `ops`
sequentially. It returns a `Promise` which resolves with the results of
each request.

**Array notation** is inspired by [LevelUP](https://github.com/Level/levelup#batch).
Each operation is an object with 3 possible properties: `type`, `key`, `value`.
`type` is either `add`, `put`, `del`, `move`, `copy`, or `clear`, and `key` is
optional when the store has a `keyPath` and the supplied value contains it.
When `move` or `copy` is used, the `key` represents the new destination of
the operation and `value` represents the old key from which to move or copy.

```js
await batch(db, 'books', [
  { type: 'add', key: 1, value: { name: 'M1', frequency: 12 } },
  { type: 'del', key: 2 }
  { type: 'put', value: { id: 3, name: 'M3', frequency: 24 } }, // no key
])
```

**Object notation** is sugar on top of array notation for `put`/`del`
operations. Set `key` to `null` in order to delete a value.

```js
await batch(db, 'storage', {
  key1: 'update value',
  key2: null, // delete value
  key3: 'new value',
})
```

### getStoreNames(storeOps: Array|Object)

Gets the store names for a set of store operations as an array.

### ConstraintError

If during sequential execution one of the operations throws a
`ConstraintError`, the `Promise` rejects with an error, but previous
successful operations will commit.
This behavior may change in future versions, as I figure out how to
properly abort transactions in `IndexedDBShim`.

## LICENSE

[MIT](./LICENSE)


History.md
# idb-batch

## ?

* Breaking change (minor): Use `sync-promise`;
* Added `getStoreNames`
* Feature: Support `transactionalBatch` (incomplete);
* Feature: Allow preexisting transaction to be supplied to `batch()` (or `transactionalBatch()`);
* Feature: Support `move`, `copy` operations;

## 1.0.0 / 2015-12-06

* initial release :sparkles:


test/index.js
// setup PhantomJS, Webkit, IE env
import 'polyfill-function-prototype-bind'
import 'indexeddbshim'
import 'regenerator-runtime-only/runtime'
import ES6Promise from 'es6-promise'

ES6Promise.polyfill()
if (navigator.userAgent.indexOf('Trident') !== -1) {
  console.log('force IE to enable compound indexes using indexeddbshim') // eslint-disable-line
  window.shimIndexedDB.__useShim()
}

import { expect } from 'chai'
import { del, open } from 'idb-factory'
import { request } from 'idb-request'
import Schema from 'idb-schema'
import batch from '../src'

describe('idb-batch', () => {
  let db
  const dbName = 'idb-batch'
  const schema = new Schema()
  .addStore('books')
  .addIndex('byTitle', 'title', { unique: true }) // simple index
  .addIndex('byAuthor', 'author')
  .addStore('magazines', { key: 'id', increment: true })
  .addIndex('byName', 'name')
  .addIndex('byNameAndFrequency', ['name', 'frequency'], { unique: true }) // compound index
  .addStore('storage')
  .addIndex('byFoo', 'foo')

  beforeEach(async () => {
    db = await open(dbName, schema.version(), schema.callback())
  })

  before(() => del(db || dbName))
  afterEach(() => del(db || dbName))

  it('supports object syntax', async () => {
    const res1 = await batch(db, 'books', {
      key1: { title: 'B1', author: 'Bob' },
      key2: { title: 'B2', author: 'Bob' },
      3: { title: 'B3', author: 'Karl' },
    })

    expect(res1.sort()).eql(['key1', 'key2', '3'].sort()) // object keys don't gurante order
    expect(await count(db, 'books')).equal(3)
    expect(await get(db, 'books', 'key1')).eql({ title: 'B1', author: 'Bob' })
    expect(await get(db, 'books', '3')).eql({ title: 'B3', author: 'Karl' })
    expect(await get(db, 'books', 3)).eql(undefined)

    const res2 = await batch(db, 'books', {
      key1: null,
      key2: null,
      3: { title: 'B3', author: 'Bob' },
    })

    expect(res2.sort()).eql([undefined, undefined, '3'].sort())
    expect(await count(db, 'books')).equal(1)
    expect(await get(db, 'books', '3')).eql({ title: 'B3', author: 'Bob' })
  })

  it('supports array syntax', async () => {
    const res1 = await batch(db, 'magazines', [
      { type: 'add', key: 1, value: { name: 'M1', frequency: 12 } },
      { type: 'add', key: 2, value: { name: 'M2', frequency: 24 } },
      { type: 'add', value: { id: 3, name: 'M3', frequency: 6 } },
      { type: 'add', value: { id: 4, name: 'M4', frequency: 52 } },
    ])

    expect(res1).eql([1, 2, 3, 4])
    expect(await count(db, 'magazines')).equal(4)
    expect(await get(db, 'magazines', 2)).eql({ id: 2, name: 'M2', frequency: 24 })
    expect(await get(db, 'magazines', 4)).eql({ id: 4, name: 'M4', frequency: 52 })

    const res2 = await batch(db, 'magazines', [
      { type: 'del', key: 1 },
      { type: 'put', key: 2, value: { name: 'M2', frequency: 24, foo: 'bar' } },
      { type: 'del', key: 3 },
    ])

    expect(res2).eql([undefined, 2, undefined])
    expect(await count(db, 'magazines')).equal(2)
    expect(await get(db, 'magazines', 2)).eql({ id: 2, name: 'M2', frequency: 24, foo: 'bar' })

    const res3 = await batch(db, 'magazines', [{ type: 'clear' }])
    expect(res3).eql([undefined])
    expect(await count(db, 'magazines')).equal(0)
  })

  it('works with any type of data (not only objects)', async () => {
    await batch(db, 'storage', {
      key1: 'value',
      key2: 123456,
      key3: [1, 2, 3],
      key4: { foo: 'bar' },
    })

    expect(await count(db, 'storage')).equal(4)
    expect(await get(db, 'storage', 'key1')).equal('value')
    expect(await get(db, 'storage', 'key2')).equal(123456)
    expect(await get(db, 'storage', 'key3')).eql([1, 2, 3])
    expect(await get(db, 'storage', 'key4')).eql({ foo: 'bar' })
  })

  it('validates unique indexes', async () => {
    const errors = []

    try {
      await batch(db, 'books', {
        1: { title: 'book', author: 'Petr' },
        2: { title: 'book', author: 'John' }, // error byTitle index
        3: { title: 'my book', author: 'John' },
      })
    } catch (err) {
      errors.push('simple index')
    }

    try {
      await batch(db, 'magazines', [
        { type: 'add', value: { name: 'magazine', frequency: 1 } },
        { type: 'add', value: { name: 'magazine', frequency: 1 } }, // error byNameAndFrequency index
        { type: 'add', value: { name: 'magazine', frequency: 2 } },
      ])
    } catch (err) {
      errors.push('compound index')
    }

    expect(errors).eql(['simple index', 'compound index'])
    expect(await count(db, 'books')).equal(1) // because transaction wasn't aborted
    expect(await count(db, 'magazines')).equal(1)
  })

  it('validates arguments', () => {
    expect(() => batch(db, 'books')).throws('invalid arguments length')
    expect(() => batch(db, 123, { a: 1 })).throws('invalid "storeName"')
    expect(() => batch(db, 'books', JSON.stringify({ a: 1 }))).throws('invalid "ops"')

    expect(() => batch(db, 'magazines', [{ type: 'delete', key: 'foo' }])).throws('invalid type "delete"')
    expect(() => batch(db, 'magazines', [['put', '1']])).throws('invalid op')
  })
})

/**
 * IDB store helpers.
 */

function count(db, storeName) {
  return request(db.transaction(storeName).objectStore(storeName).count())
}

function get(db, storeName, key) {
  return request(db.transaction(storeName).objectStore(storeName).get(key))
}


src/index.js
import 'babel-polyfill' // Array.prototype.values, etc.

import isPlainObj from 'is-plain-obj'
import isSafari from 'is-safari'
import SyncPromise from 'sync-promise'

/**
 * Links to array prototype methods.
 */

const slice = [].slice
const map = [].map

/**
 * Perform batch operation for a single object store using `ops`.
 *
 * Array syntax:
 *
 * [
 *   { type: 'add', key: 'key1', val: 'val1' },
 *   { type: 'put', key: 'key2', val: 'val2' },
 *   { type: 'del', key: 'key3' },
 * ]
 *
 * Object syntax:
 *
 * {
 * 	 key1: 'val1', // put val1 to key1
 * 	 key2: 'val2', // put val2 to key2
 * 	 key3: null,   // delete key
 * }
 *
 * @param {Array|Object} ops Operations
 * @param {Object} opts See `transactionalBatch`
 * @return {Promise} Resolves to the result of the operations
 */

export default function batch(db, storeName, ops, opts) {
  if (typeof storeName !== 'string') throw new TypeError('invalid "storeName"')
  if (![3, 4].includes(arguments.length)) throw new TypeError('invalid arguments length')
  validateAndCanonicalizeOps(ops)
  return transactionalBatch(db, { [storeName]: ops }, opts).then((arr) => arr[0].map((obj) => obj[storeName]))
}

export function getStoreNames(storeOpsArr) {
  if (isPlainObj(storeOpsArr)) storeOpsArr = [storeOpsArr]
  return storeOpsArr.reduce((storeNames, opObj) => {
    // This has no effect if the opObj is a function
    return storeNames.concat(Object.keys(opObj))
  }, [])
}

/**
 * Perform batch operations for any number of object stores using `ops`.
 *
 * Array syntax:
 *
 * [
 *   { type: 'add', key: 'key1', val: 'val1' },
 *   { type: 'put', key: 'key2', val: 'val2' },
 *   { type: 'del', key: 'key3' },
 * ]
 *
 * Object syntax:
 *
 * {
 * 	 key1: 'val1', // put val1 to key1
 * 	 key2: 'val2', // put val2 to key2
 * 	 key3: null,   // delete key
 * }
 *
 * @param {IDBDatabase|IDBTransaction} tr IndexedDB database or transaction on which the batch will operate
 * @param {Array|Object} storeOpsArr Array of objects (or a single object) whose keys are store names and objects are idb-batch operations (object or array)
 * @param {Object} [opts] Options object
 * @property {Boolean} opts.parallel=false Whether or not to load in parallel
 * @property {Array} opts.extraStores=[]] A list of store names to add to the transaction (when `tr` is an `IDBDatabase` object)
 * @return {Promise} Resolves to an array containing the results of the operations for each store
 */

export function transactionalBatch(tr, storeOpsArr, opts = { parallel: false, extraStores: [] }) {
  if (![2, 3].includes(arguments.length)) throw new TypeError('invalid arguments length')
  if (isPlainObj(storeOpsArr)) storeOpsArr = [storeOpsArr]
  const storeOpsArrIter = storeOpsArr.values()
  if (typeof tr.createObjectStore === 'function') {
    tr = tr.transaction(getStoreNames(storeOpsArr).concat(opts.extraStores), 'readwrite')
  }
  return new SyncPromise((resolve, reject) => {
    const results = []
    tr.onerror = handleError(reject)
    tr.oncomplete = () => resolve(results)

    if (opts.parallel) {
      return
    }

    let iterateStoreOps
    const iterateStores = (storeOpsVals, storeOpsObj) => {
      if (typeof storeOpsObj === 'function') {
        storeOpsObj(tr)
        iterateStoreOps()
        return
      }
      const storeOpIter = storeOpsVals.next()
      if (storeOpIter.done) {
        iterateStoreOps()
        return
      }
      const storeName = storeOpIter.value
      const storeResults = []

      let ops = storeOpsObj[storeName]
      try {
        ops = validateAndCanonicalizeOps(ops)
      } catch (err) {
        reject(err)
        return
      }

      const store = tr.objectStore(storeName)
      let currentIndex = 0

      next()

      function next() {
        const { type, key } = ops[currentIndex]
        if (type === 'clear') {
          request(storeName, store.clear())
          return
        }
        if (type === 'del') {
          request(storeName, store.delete(key))
          return
        }
        const val = ops[currentIndex].val || ops[currentIndex].value
        if (['move', 'copy'].includes(type)) {
          const req = store.get(val)
          req.onerror = handleError(reject)
          req.onsuccess = (e) => {
            ops.splice(currentIndex, 0, { type: 'put', key, value: e.target.result })
            if (type === 'move') {
              ops.splice(currentIndex + 1, 0, { type: 'delete', key: val })
            }
            next()
          }
          return
        }
        if (key && store.keyPath) val[store.keyPath] = key

        countUniqueIndexes(store, key, val, (err, uniqueRecordsCounter) => {
          if (err) return reject(err)

          // We don't abort transaction here (we just stop execution)
          // Browsers' implementations also don't abort, and instead just throw an error
          if (uniqueRecordsCounter) return reject(new Error('Unique index ConstraintError'))
          request(storeName, store.keyPath ? store[type](val) : store[type](val, key))
        })
      }

      function request(storeNm, req) {
        currentIndex++

        req.onerror = handleError(reject)
        req.onsuccess = (e) => {
          storeResults.push({ [storeNm]: e.target.result })
          if (currentIndex < ops.length) next()
          else {
            results.push(storeResults)
            iterateStores(storeOpsVals, storeOpsObj)
          }
        }
      }
    }

    iterateStoreOps = () => {
      let storeOpsObj = storeOpsArrIter.next()
      if (storeOpsObj.done) {
        return
      }
      storeOpsObj = storeOpsObj.value
      const storeOpsVals = Object.keys(storeOpsObj).values()
      iterateStores(storeOpsVals, storeOpsObj)
    }
    iterateStoreOps()
  })
}

/**
 * Validate unique index manually.
 *
 * Fixing:
 * - https://bugs.webkit.org/show_bug.cgi?id=149107
 * - https://github.com/axemclion/IndexedDBShim/issues/56
 *
 * @param {IDBStore} store
 * @param {Any} val
 * @param {Function} cb(err, uniqueRecordsCounter)
 */

function countUniqueIndexes(store, key, val, cb) {
  // rely on native support
  if (!isSafari && global.indexedDB !== global.shimIndexedDB) return cb()

  const indexes = slice.call(store.indexNames).map((indexName) => {
    const index = store.index(indexName)
    const indexVal = isCompound(index)
    ? map.call(index.keyPath, (indexKey) => val[indexKey]).filter((v) => v)
    : val[index.keyPath]

    return [index, indexVal]
  }).filter(([index, indexVal]) => {
    return index.unique && (isCompound(index) ? indexVal.length : indexVal)
  })

  if (!indexes.length) return cb()

  let totalRequestsCounter = indexes.length
  let uniqueRecordsCounter = 0

  indexes.forEach(([index, indexVal]) => {
    const req = index.getKey(indexVal) // get primaryKey to compare with updating value
    req.onerror = handleError(cb)
    req.onsuccess = (e) => {
      if (e.target.result && e.target.result !== key) uniqueRecordsCounter++
      totalRequestsCounter--
      if (totalRequestsCounter === 0) cb(null, uniqueRecordsCounter)
    }
  })
}

/**
 * Check if `index` is compound
 *
 * @param {IDBIndex} index
 * @return {Boolean}
 */

function isCompound(index) {
  return typeof index.keyPath !== 'string'
}

/**
 * Create error handler.
 *
 * @param {Function} cb
 * @return {Function}
 */

function handleError(cb) {
  return (e) => {
    // prevent global error throw https://bugzilla.mozilla.org/show_bug.cgi?id=872873
    if (typeof e.preventDefault === 'function') e.preventDefault()
    cb(e.target.error)
  }
}

/**
 * Validate operations and canonicalize them into an array.
 *
 * @param {Array|Object} ops
 * @return {Array} Canonicalized operations array
 */

function validateAndCanonicalizeOps(ops) {
  if (ops === 'clear') {
    ops = [{ type: 'clear' }]
  }
  if (!Array.isArray(ops) && !isPlainObj(ops)) {
    throw new TypeError('invalid "ops"')
  }
  if (isPlainObj(ops)) {
    ops = Object.keys(ops).map((key) => {
      return { key, value: ops[key], type: ops[key] === null ? 'del' : 'put' }
    })
  }
  ops.forEach((op) => {
    if (!isPlainObj(op)) throw new TypeError('invalid op')
    if (['add', 'put', 'del', 'move', 'copy', 'clear'].indexOf(op.type) === -1) throw new TypeError(`invalid type "${op.type}"`)
  })
  return ops
}
